tests/quickcheck.rs

```diff
@@ -0,0 +1,326 @@
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_sub_primitive(a: u128, b: u128) -> bool {
+            if b < a {
+                BigUint::from(a-b) == BigUint::from(a) - BigUint::from(b)
+            } else {
+                BigUint::from(b-a) == BigUint::from(b) - BigUint::from(a)
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_sub_primitive(a: i128, b: i128) -> bool {
+            if b < a {
+                BigInt::from(a-b) == BigInt::from(a) - BigInt::from(b)
+            } else {
+                BigInt::from(b-a) == BigInt::from(b) - BigInt::from(a)
+            }
+        }
+    }
+
+    quickcheck! {
+        /// Tests correctness of subtraction assuming addition is correct
+        fn quickcheck_unsigned_sub(a: BigUint, b: BigUint) -> bool {
+            if b < a {
+                a.clone() - b.clone() + b == a
+            } else {
+                b.clone() - a.clone() + a == b
+            }
+        }
+    }
+
+    quickcheck! {
+        /// Tests correctness of subtraction assuming addition is correct
+        fn quickcheck_signed_sub(a: BigInt, b: BigInt) -> bool {
+            if b < a {
+                a.clone() - b.clone() + b == a
+            } else {
+                b.clone() - a.clone() + a == b
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_div_primitive(a: u128, b: u128) -> TestResult {
+            if b == 0 {
+                TestResult::discard()
+            } else {
+                TestResult::from_bool(BigUint::from(a/b) == BigUint::from(a) / BigUint::from(b))
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_div_primitive(a: i128, b: i128) -> TestResult {
+            if b == 0 {
+                TestResult::discard()
+            } else {
+                TestResult::from_bool(BigInt::from(a/b) == BigInt::from(a) / BigInt::from(b))
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_pow_zero(a: BigUint) -> bool {
+            a.pow(0_u32) == BigUint::one()
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_pow_one(a: BigUint) -> bool {
+            a.pow(1_u32) == a
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_sqrt(a: BigUint) -> bool {
+            (a.clone() * a.clone()).sqrt() == a
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_cbrt(a: BigUint) -> bool {
+            (a.clone() * a.clone() * a.clone()).cbrt() == a
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_cbrt(a: BigInt) -> bool {
+            (a.clone() * a.clone() * a.clone()).cbrt() == a
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_conversion(a:BigUint) -> bool {
+            let mut success = true;
+            for radix in 2..=36 {
+                let string = a.to_str_radix(radix);
+                success &= a == BigUint::from_str_radix(&string, radix).unwrap()
```

Comment:
I'll consider removing this function, as it takes ~5s to run on my computer

Reply:
maybe restrict radix to fewer values - 2,4,8,10,16,32?

Reply:
Maybe let the radix be another quickcheck-provided argument? I guess you can discard those outside of `2..=36`, or maybe there's a better way to deal with that...

Reply:
radix is now a quickcheck provided parameter. There's a slight problem in that ~65% of tests will be discarded. In this case I will leave as-is, but I could additionally hack on a `let radix = radix % 34 + 2`. That would make the test less readable (quickcheck might say it failed for radix = 100, but really radix is 100 % 34 + 2= 34) but the test would never be discarded.

Reply:
Should be mod 35 if we want to reach the maximum 36 -- `(0..35) + 2 = (2..37) = (2..=36)`. Or perhaps you could use `StdThreadGen` for a custom radix range?

Reply:
Whoops I got the formula wrong. I was just spit-balling though, that isn't in code.

Here's what the check looks like now:
```
    if radix > 36 || radix < 2 {
        return TestResult::discard();
    }
```
I think this is fine and I'd prefer to leave it this way, but in other situations, it might not be fine (like if the range were looking for was 0-3, as that is too low a percentage). Quickcheck aims for a certain number of tests to pass (i.e. not be discarded), but it has a maximum number of tests it will try.

> By default, if quickcheck can't find 100 valid tests after trying 10,000 times, then it will give up.

So there is no problem with the current implementation, which discards ~2/3 of tests.

I don't thing you could use a `StdThreadGen` to limit the range because quickcheck uses the same `StdThreadGen` for all variables in the function. So then we would end up with only really small numbers that are being converted.